File: backend/internal/sales/domain/returns.go
package domain

import (
	"errors"
	"time"

	"github.com/google/uuid"
)

var ErrReturnWindowExpired = errors.New("prazo para devolução expirado")

type CustomerReturn struct {
	ID                uuid.UUID            `json:"id"`
	ReturnNumber      string               `json:"return_number"`
	OrderID           uuid.UUID            `json:"order_id"`
	OrderNumber       string               `json:"order_number,omitempty"`
	CustomerID        uuid.UUID            `json:"customer_id"`
	CustomerName      string               `json:"customer_name,omitempty"`
	Status            string               `json:"status"`
	Reason            string               `json:"reason"`
	ConditionNotes    *string              `json:"condition_notes,omitempty"`
	ReturnWindowDays  int                  `json:"return_window_days,omitempty"`
	ReturnExpiresAt   *time.Time           `json:"return_expires_at,omitempty"`
	WithinReturnWindow bool                `json:"within_return_window"`
	Resolution        *string              `json:"resolution,omitempty"`
	Notes             *string              `json:"notes,omitempty"`
	RequestedBy       *uuid.UUID           `json:"requested_by,omitempty"`
	ApprovedBy        *uuid.UUID           `json:"approved_by,omitempty"`
	Items             []CustomerReturnItem `json:"items,omitempty"`
	Photos            []CustomerReturnPhoto `json:"photos,omitempty"`
	CreatedAt         time.Time            `json:"created_at"`
	UpdatedAt         time.Time            `json:"updated_at"`
}

type CustomerReturnPhoto struct {
	ID        uuid.UUID `json:"id"`
	ReturnID  uuid.UUID `json:"return_id"`
	CreatedAt time.Time `json:"created_at"`
}

type CustomerReturnItem struct {
	ID              uuid.UUID  `json:"id"`
	ReturnID        uuid.UUID  `json:"return_id"`
	OrderItemID     *uuid.UUID `json:"order_item_id,omitempty"`
	SKUID           uuid.UUID  `json:"sku_id"`
	SKUCode         string     `json:"sku_code,omitempty"`
	InventoryUnitID *uuid.UUID `json:"inventory_unit_id,omitempty"`
	Quantity        int        `json:"quantity"`
	ConditionNotes  *string    `json:"condition_notes,omitempty"`
}

type ReturnPhotoUpload struct {
	Body []byte
	Ext  string
}

type CreateCustomerReturnInput struct {
	OrderID        uuid.UUID
	Reason         string
	ConditionNotes *string
	Notes          *string
	RequestedBy    uuid.UUID
	Items          []CreateReturnItemInput
	Photos         []ReturnPhotoUpload
}

type CreateReturnItemInput struct {
	OrderItemID     *uuid.UUID `json:"order_item_id,omitempty"`
	SKUID           uuid.UUID  `json:"sku_id"`
	InventoryUnitID *uuid.UUID `json:"inventory_unit_id,omitempty"`
	Quantity        int        `json:"quantity"`
	ConditionNotes  *string    `json:"condition_notes,omitempty"`
}

File: admin/lib/api/returns.ts
/**
 * @file returns.ts
 * @description /api/v1/sales/returns — devoluções comerciais.
 * @hooks hooks: use-customer-returns-*, use-return-step
 *
 * @see admin/lib/api/README.md
 */

import { api, apiBlob, apiForm } from "./client";
import type { Order, OrderItem, OrderListItem } from "../types";
import { salesApi } from "./sales";

const BASE = "/api/v1/sales/returns";

export type ReturnPhoto = {
  id: string;
  return_id: string;
  created_at: string;
};

export type CustomerReturn = {
  id: string;
  return_number: string;
  order_number?: string;
  customer_name?: string;
  status: string;
  reason: string;
  condition_notes?: string;
  within_return_window: boolean;
  return_window_days?: number;
  return_expires_at?: string;
  photos?: ReturnPhoto[];
  resolution?: string;
  created_at: string;
};

export type ReturnWindowCheck = {
  return_window_days: number;
  return_expires_at?: string;
  within_return_window: boolean;
};

export const returnsApi = {
  list: (q?: string) => {
    const qs = q?.trim() ? `?q=${encodeURIComponent(q.trim())}` : "";
    return api<{ items: CustomerReturn[] }>(`${BASE}${qs}`);
  },
  get: (id: string) => api<CustomerReturn>(`${BASE}/${id}`),
  windowCheck: (orderId: string) =>
    api<ReturnWindowCheck>(`${BASE}/window-check?order_id=${encodeURIComponent(orderId)}`),
  eligibility: (orderId: string, orderItemId: string) =>
    api<{ eligible_units: number }>(
      `${BASE}/eligibility?order_id=${encodeURIComponent(orderId)}&order_item_id=${encodeURIComponent(orderItemId)}`,
    ),
  loadOrderContext: async (orderId: string) => {
    const [order, windowInfo] = await Promise.all([
      salesApi.getOrder(orderId),
      returnsApi.windowCheck(orderId),
    ]);
    return { order, orderItems: (order.items ?? []) as OrderItem[], windowInfo };
  },
  searchShippedOrders: (q: string) => salesApi.searchShippedOrders(q),
  createWithPhotos: (form: FormData) => apiForm(`${BASE}`, form),
  photoBlob: (returnId: string, photoId: string) =>
    apiBlob(`${BASE}/${returnId}/photos/${photoId}/file`),
  step: (id: string, step: string, body?: Record<string, unknown>) =>
    api(`${BASE}/${id}/${step}`, {
      method: "POST",
      body: body ? JSON.stringify(body) : undefined,
    }),
};

export type { Order, OrderListItem };
